notebooks/advanced/Clustering_Preprocessing/01. Optional - Data Cleaning and Preparation.ipynb

```diff
@@ -0,0 +1,1403 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## (Optional) Data Cleaning and Preparation\n",
+    "\n",
+    "This notebook pulls the UCI Online Retail II Data Set and cleans / preprocesses it for use with clustering algorithms.\n",
+    "\n",
+    "Data Source: https://archive.ics.uci.edu/ml/datasets/Online+Retail+II\n",
+    "\n",
+    "File Location: https://archive.ics.uci.edu/ml/machine-learning-databases/00502/"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import os\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "from urllib.request import urlretrieve\n",
+    "\n",
+    "import warnings\n",
+    "warnings.filterwarnings(\"ignore\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "('./data/online_retail_II.xlsx', <http.client.HTTPMessage at 0x1c145029190>)"
+      ]
+     },
+     "execution_count": 2,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# pull data file from source\n",
+    "dir_path = './data'\n",
+    "\n",
+    "if not os.path.exists(dir_path):\n",
+    "    os.makedirs(dir_path)\n",
+    "    \n",
+    "urlretrieve(\"https://archive.ics.uci.edu/ml/machine-learning-databases/00502/online_retail_II.xlsx\",\n",
+    "            f\"{dir_path}/online_retail_II.xlsx\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%pip install openpyxl"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
```

Comment:
Two notebooks contain output cells. If it is not intentional, please clear output cells.

Reply:
Cleaned, thanks